### ISSUE TITLE
TODO

### ISSUE THREAD

**Issue by EricChen424** [bcgsc/GapPredict#33]
1) Extend to predict d > 1 bases and see how it performs
- **"Done"**... the statistics should now be pretty good for d > 1 base but I need to hook this up with the Keras model and test it
2) Try making the training set an actual chromosome (or even just a tiny chunk of a contiguous region so our kmers actually have some context)
- **"Done"**... we have a shell script to extract all reads mapping to a contiguous region of an actual bacterial chromosome now
3) Train on an entire genome, start at some random kmer and estimate the next d bases
- plot the base correctness against how many bases down you're predicting and see when we get to 25%
- **Not really done**... I've tried this on a contiguous stretch of reads and got very good accuracy but this is only for a 1 Kb stretch (which isn't very long)... also this is only next 1 base prediction and haven't expanded this to d bases yet
4) Take an entire genome, delete some chunks out of it, select some seeds at the edge of those chunks and see if you can predict the next d unknown kmers
- we will train on the deleted chunks
- **Done**, or at least the skeleton to predict d unknown bases given a model predicted on any number of bases

Shannon entropy?
ntHits?
cBF - tries to keep the kmers that are well covered (and thus not likely to be incorrect) but also not too redundant (and thus likely have a poor mapping)
- use this to filter out reads
-----------------
**Nov 7**
First should make sure that my "hole" in the dataset dataset is actually a valid hole.
Then probably want to look into hooking up the model saving functionality so we can hook up an application that predicts an arbitrary number of bases 
- Note: It seems to make a bit less sense to be predicting say 30 bases if we only trained on the next 1 base... although it would be interesting to see how much accuracy falls off; however we can compare training on the next 1 base and the next 30 bases
---------------
Scalability - make this run on terminal (to see if we can use the stronger hardware), add more technology, can we use C++?
How does increasing the k of k-mers affect the prediction and training time?
Get a deeper understanding of the architecture of this model and whether it's appropriate for our task?
- Does a CNN make sense for this problem?

-----------------
Perhaps e-mail the MachineLearningMastery guy and ask him if there are ways to optimize the model
**Use the French-English prediction as a benchmark
Look into Google machine learning (eg. their autocomplete, their sentence completion models too)?**
Maybe just other sentence completion models as well
See if we can find benchmark data for other RNN/LSTM models (and also with different frameworks) - make sure it's working on big data as well

https://github.com/tensorflow/nmt

**Also, hook up the command line stuff, run on a 10k bp contig, get the runtime stats**

--------------------
Nov 21
Parameter sweep
- check how different hyperparameters affect runtime and accuracy

Consider:
1) **A vanilla neural network**
- Encode your read (eg. one hot), predict the next base 
- flatten it (4xk vector)
- try both onehot encoding and "w-mer" embedding (w << k)
**- consider a conv net (take say 4x w windows with w << k, sliding window)**
- take in a matrix and output a base
- one hot encoding
2) Kmer embeddings (not just 1 hot encodings)
3) Run ntHits and filter out kmers (or the count based filtering)
- gets rid of low confidence kmers
4) See how coverage affects model quality:
- eg. sample 10% to simulate say 6X coverage for a 60X coverage 
5) Get a set of contigs and a set of reads not represented in the assembly (this set of reads might be small enough)
- train on the reads and see if we can patch the contigs/scaffold
6) See if we can increase the number of threads in HPCE
7)
